rust/agents/relayer/src/msg/serial_submitter.rs

```diff
@@ -0,0 +1,232 @@
+/// The reasons a message might not be eligible are:
+///
+///  *  Insufficient interchain gas payment on source chain
+///  *  Already delivered to destination chain, e.g. maybe by a different relayer, or the result of
+///     a submission attempt just prior to an old incarnation of this task crashing.
+///  *  Not whitelisted (currently checked by processor)
+///  *  Wrong destination chain (currently checked by processor)
+///  *  Checkpoint index < leaf index (currently checked by processor)
+///
+// Therefore, we maintain two queues of messages:
+///
+///   1.  run_queue: messages which are eligible for submission but waiting for
+///       their turn to run, since we can only do one at a time.
+///
+///   2.  wait_queue: messages currently ineligible for submission, due to one of the
+///       reasons listed above (e.g. index not covered by checkpoint, insufficient gas, etc).
+///
+/// Note that there is no retry queue. This is because if submission fails for a retriable
+/// reason, the message instead goes directly back on to the runnable queue (though it will be
+/// prioritized lower than it was prior to the failed attempt due to the increased
+/// num_retries).
+///
+/// To summarize: each scheduler `tick()`, new messages from the processor are inserted onto
+/// the wait queue.  We then scan the wait_queue, looking for messages which can be promoted to
+/// the runnable_queue, e.g. by comparing with a recent checkpoint or latest gas payments on
+/// source chain. If eligible for delivery, the message is promoted to the runnable queue and
+/// prioritized accordingly. Note that for messages that have never been attempted before, they
+/// will sort very highly due to num_retries==0 and probably be tried soon.
+
+// TODO(webbhorn): Metrics data.
+// TODO(webbhorn): Do we also want to await finality_blocks on source chain before attempting
+// submission? Does this already happen?
+
+#[allow(dead_code)]
+#[derive(Debug)]
+pub(crate) struct SerialSubmitter {
+    // Receiver for new messages to submit.
+    rx: mpsc::Receiver<SubmitMessageArgs>,
+    // Messages we are aware of that we want to eventually submit, but haven't yet, for
+    // whatever reason. They are not in any priority order, so are held in a vector.
+    wait_queue: Vec<SubmitMessageArgs>,
+    // Messages that are in theory deliverable, but which are waiting in a queue for their turn
+    // to be dispatched. The SerialSubmitter can only dispatch one message at a time, so this
+    // queue could grow.
+    run_queue: BinaryHeap<SubmitMessageArgs>,
+    // Inbox / InboxValidatorManager on the destination chain.
+    inbox_contracts: InboxContracts,
+    // Outbox on message origin chain.
+    outbox: Outboxes,
+    // Contract tracking interchain gas payments for use when deciding whether sufficient funds
+    // have been provided for message forwarding.
+    interchain_gas_paymaster: Option<Arc<CachingInterchainGasPaymaster>>,
```

Comment:
I don't think we actually need this, at least not in the current way we have the code set up. To save gas, the contract on chain doesn't store the payments made for a particular message, instead we index events emitted by the contract and store the payments locally in `AbacusDB`. The way to get total payment made for a message is (for better or worse) by calling this function in AbacusDB: https://github.com/abacus-network/abacus-monorepo/blob/main/rust/abacus-core/src/db/abacus_db.rs#L292, so I think we can remove this

Reply:
Done, thanks, I've dropped IGP from submitters.